.NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateParser.cs

```diff
@@ -403,11 +403,100 @@ private DateTimeResolutionResult ParseImplicitDate(string text, DateObject refer
                 return ret;
             }
 
+            // Handling cases like 'Monday 21', which both 'Monday' and '21' refer to the same date.
+            // The year of expected date can be different to the year of referenceDate.
+            match = this.config.WeekDayAndDayRegex.Match(text);
+            if (match.Success)
+            {
+                int month = referenceDate.Month, year = referenceDate.Year;
+
+                // Create a extract result which content ordinal string of text
+                ExtractResult ertmp = new ExtractResult
+                {
+                    Text = match.Groups["day"].Value,
+                    Start = match.Groups["day"].Index,
+                    Length = match.Groups["day"].Length,
+                };
+
+                // Parse the day in text into number
+                var day = Convert.ToInt32((double)(this.config.NumberParser.Parse(ertmp).Value ?? 0));
+
+                var pivotDate = new DateObject(year, month, 1);
+                var daysInMonth = DateObject.DaysInMonth(year, month);
+                if (daysInMonth >= day)
+                {
+                    pivotDate = DateObject.MinValue.SafeCreateFromValue(year, month, day);
+                }
+                else
+                {
+                    // Add 1 month is enough, since 1, 3, 5, 6, 8, 10, 12 months has 31 days
+                    pivotDate = pivotDate.AddMonths(1);
+                    pivotDate = DateObject.MinValue.SafeCreateFromValue(pivotDate.Year, pivotDate.Month, day);
+                }
+
+                var numWeekDayInt = (int)pivotDate.DayOfWeek;
+                var extractedWeekDayStr = match.Groups["weekday"].Value.ToLower();
+                var weekDay = this.config.DayOfWeek[extractedWeekDayStr];
+                if (!pivotDate.Equals(DateObject.MinValue))
+                {
+                    if (day == referenceDate.Day && numWeekDayInt == weekDay)
+                    {
+                        ret.FutureValue = new DateObject(year, month, day);
+                        ret.PastValue = new DateObject(year, month, day);
+                        ret.Timex = DateTimeFormatUtil.LuisDate(year, month, day);
+                    }
+                    else
+                    {
+                        var futureDate = pivotDate;
+                        var pastDate = pivotDate;
+
+                        while ((int)futureDate.DayOfWeek != weekDay || futureDate.Day != day || futureDate < referenceDate)
+                        {
+                            futureDate = futureDate.AddMonths(1);
```

Comment:
I just made an experiment to confirm the behavior of AddMonths.
The implementation here is correct but it may lead to some confusion. Please add some comments to make the behavior even clear.
Thanks.